File: ios/Gogai/Repositories/ArticleRepository.swift
import Foundation

struct ArticleRepository: Sendable {
    private let client: any APIClientProtocol

    init(client: any APIClientProtocol) {
        self.client = client
    }

    func fetchAll(feedId: Int? = nil, groupId: Int? = nil, filter: ArticleFilter = .all, sortOrder: ArticleSortOrder = .publishedAt, limit: Int = 1000, offset: Int = 0, includeSecret: Bool = false) async throws -> [Article] {
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "limit", value: String(limit)),
            URLQueryItem(name: "offset", value: String(offset)),
            URLQueryItem(name: "sortBy", value: sortOrder.rawValue),
        ]
        if let feedId { queryItems.append(URLQueryItem(name: "feedId", value: String(feedId))) }
        if let groupId { queryItems.append(URLQueryItem(name: "groupId", value: String(groupId))) }
        switch filter {
        case .all: break
        case .unread: queryItems.append(URLQueryItem(name: "unreadOnly", value: "true"))
        // likedOnly 指定時はサーバーが sortBy を無視して liked_at 降順で返す
        case .liked: queryItems.append(URLQueryItem(name: "likedOnly", value: "true"))
        }
        if includeSecret { queryItems.append(URLQueryItem(name: "includeSecret", value: "true")) }
        return try await client.send(.get("/api/articles", queryItems: queryItems))
    }

    func fetchCounts() async throws -> [FeedCount] {
        try await client.send(.get("/api/articles/counts"))
    }

    func fetch(id: Int) async throws -> Article {
        try await client.send(.get("/api/articles/\(id)"))
    }

    func markAsRead(id: Int) async throws {
        try await client.sendVoid(.post("/api/articles/\(id)/read"))
    }

    func markAsUnread(id: Int) async throws {
        try await client.sendVoid(.post("/api/articles/\(id)/unread"))
    }

    func like(id: Int) async throws {
        try await client.sendVoid(.post("/api/articles/\(id)/like"))
    }

    func unlike(id: Int) async throws {
        try await client.sendVoid(.post("/api/articles/\(id)/unlike"))
    }
}

File: android/app/src/main/kotlin/com/mtkg/gogai/repository/ArticleRepository.kt
package com.mtkg.gogai.repository

import com.mtkg.gogai.model.Article
import com.mtkg.gogai.model.ArticleFilter
import com.mtkg.gogai.model.ArticleSortOrder
import com.mtkg.gogai.model.FeedCount
import com.mtkg.gogai.network.ApiClient
import com.mtkg.gogai.network.Endpoint
import kotlinx.serialization.builtins.ListSerializer

/// 記事 API（iOS ArticleRepository の移植）
class ArticleRepository(private val client: ApiClient) {

    suspend fun fetchAll(
        feedId: Int? = null,
        groupId: Int? = null,
        filter: ArticleFilter = ArticleFilter.All,
        sortOrder: ArticleSortOrder = ArticleSortOrder.PublishedAt,
        limit: Int = 1000,
        offset: Int = 0,
        includeSecret: Boolean = false,
    ): List<Article> {
        val queryItems = buildList {
            add("limit" to limit.toString())
            add("offset" to offset.toString())
            add("sortBy" to sortOrder.rawValue)
            feedId?.let { add("feedId" to it.toString()) }
            groupId?.let { add("groupId" to it.toString()) }
            when (filter) {
                ArticleFilter.All -> Unit
                ArticleFilter.Unread -> add("unreadOnly" to "true")
                // likedOnly 指定時はサーバーが sortBy を無視して liked_at 降順で返す
                ArticleFilter.Liked -> add("likedOnly" to "true")
            }
            if (includeSecret) add("includeSecret" to "true")
        }
        return client.send(Endpoint.get("/api/articles", queryItems), ListSerializer(Article.serializer()))
    }

    suspend fun fetchCounts(): List<FeedCount> =
        client.send(Endpoint.get("/api/articles/counts"), ListSerializer(FeedCount.serializer()))

    suspend fun fetch(id: Int): Article =
        client.send(Endpoint.get("/api/articles/$id"), Article.serializer())

    suspend fun markAsRead(id: Int) {
        client.sendVoid(Endpoint.post("/api/articles/$id/read"))
    }

    suspend fun markAsUnread(id: Int) {
        client.sendVoid(Endpoint.post("/api/articles/$id/unread"))
    }

    suspend fun like(id: Int) {
        client.sendVoid(Endpoint.post("/api/articles/$id/like"))
    }

    suspend fun unlike(id: Int) {
        client.sendVoid(Endpoint.post("/api/articles/$id/unlike"))
    }
}
